### PR TITLE
[vcr-2.0] Fix nullptr in helix path

### PR DESCRIPTION
**Pull request by snalli**
* Fixes nullptr in helix path
* Also removes some un-used compaction code
* Other minor logging fixes

```
java.lang.reflect.InvocationTargetException: null
        at jdk.internal.reflect.GeneratedMethodAccessor62.invoke(Unknown Source) ~[?:?]
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
        at org.apache.helix.messaging.handling.HelixStateTransitionHandler.invoke(HelixStateTransitionHandler.java:350) ~[org.apache.helix.helix-core-1.1.0.0.jar:1.1.0.0]
        at org.apache.helix.messaging.handling.HelixStateTransitionHandler.handleMessage(HelixStateTransitionHandler.java:278) [org.apache.helix.helix-core-1.1.0.0.jar:1.1.0.0]
        at org.apache.helix.messaging.handling.HelixTask.call(HelixTask.java:97) [org.apache.helix.helix-core-1.1.0.0.jar:1.1.0.0]
        at org.apache.helix.messaging.handling.HelixTask.call(HelixTask.java:49) [org.apache.helix.helix-core-1.1.0.0.jar:1.1.0.0]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.lang.NullPointerException
        at com.github.ambry.cloud.VcrReplicationManager$1.onPartitionAdded(VcrReplicationManager.java:304) ~[com.github.ambry.ambry-cloud-0.4.452.jar:?]
```

### CODE REVIEW

File: ambry-commons/src/main/java/com/github/ambry/clustermap/HelixVcrUtil.java [hunk 1]
```diff
@@ -314,25 +309,27 @@ public static boolean isSrcDestSync(String srcZkString, String srcClusterName, S
...
+        logger.error("[HelixSync] Resource {} from source Ambry cluster is absent in VCR cluster", resource);
+        status = false;
+      }
+      // check if every partition exist.
+      Set<String> srcPartitions = srcAdmin.getResourceIdealState(srcClusterName, resource).getPartitionSet();
+      Set<String> destPartitions = destAdmin.getResourceIdealState(destClusterName, resource).getPartitionSet();
+      for (String partition : srcPartitions) {
+        if (!destPartitions.contains(partition)) {
+          logger.error("[HelixSync] Partition {} from source Ambry cluster is absent in VCR cluster", partition);
+          status = false;
```
Comment (justinlin-linkedin) [inline ambry-commons/src/main/java/com/github/ambry/clustermap/HelixVcrUtil.java:328]: We don't need to go through all the inconsistent resources and partition to return false, we can just return false when first inconsistency is detected. What is the purpose of having iterating through all of the inconsistency?

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/ConnectionStringBasedStorageClient.java [hunk 2]
```diff
@@ -105,11 +107,18 @@ protected BlobServiceClient buildBlobServiceSyncClient(HttpClient httpClient, Co
...
+    try {
+      return new TableServiceClientBuilder().connectionString(azureCloudConfig.azureTableConnectionString)
+          .httpClient(httpClient)
+          .retryOptions(retryOptions)
+          .configuration(configuration)
+          .buildClient();
+    } catch (Throwable e) {
+      String err = String.format("Failed to build table service client with connection string %s due to %s",
+          azureCloudConfig.azureTableConnectionString, e.getMessage());
+      logger.error(err);
```
Comment (justinlin-linkedin) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/azure/ConnectionStringBasedStorageClient.java:119]: logger is actually a much better string formatter than you think. 
You can do 
```
logger.error("Failed to build table service client with connection string {} due to {}",
          azureCloudConfig.azureTableConnectionString, e.getMessage());
```

File: ambry-cloud/src/main/java/com/github/ambry/cloud/VcrReplicationManager.java [hunk 3]
```diff
@@ -355,30 +353,6 @@ protected void scheduleTasks() {
...
-   * @param intervalSec period between successive executions.
-   * @param taskName name of the task being scheduled.
-   */
-  private void scheduleTask(Runnable task, boolean isEnabled, long delaySec, long intervalSec, String taskName) {
-    if (isEnabled) {
-      scheduler.scheduleAtFixedRate(task, delaySec, intervalSec, TimeUnit.SECONDS);
-      logger.info("Scheduled {} task to run every {} seconds starting in {} seconds.", taskName, intervalSec, delaySec);
-    } else {
-      logger.warn("Running with {} turned off!", taskName);
-    }
```
Comment (justinlin-linkedin) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/VcrReplicationManager.java:381]: why are we removing the container deprecation logic here. We still deletes containers and once those containers are deleted, we need cloud to clean up all the blobs that belong to those containers.

File: ambry-vcr/src/main/java/com/github/ambry/vcr/VcrServer.java [hunk 4]
```diff
@@ -222,6 +222,7 @@ public void startup() throws InstantiationException {
               clusterMapConfig, clusterMap, accountService, storeConfig, cloudDestination,
               registry)).getVcrClusterParticipant();
       cloudStorageManager = new CloudStorageManager(properties, vcrMetrics, cloudDestination, clusterMap);
+      scheduler = Utils.newScheduler(1, "vcr-scheduler-",true);
```
[the remark this answers is not in the document]
Comment (snalli) [reply ambry-vcr/src/main/java/com/github/ambry/vcr/VcrServer.java:225]: yeah the scheduler was null, so the partition sync never ran. 
however compaction has been running, so no worries there.

File: ambry-commons/src/main/java/com/github/ambry/clustermap/HelixVcrUtil.java (hunk 1 above)
Comment (snalli) [reply ambry-commons/src/main/java/com/github/ambry/clustermap/HelixVcrUtil.java:328]: the idea is to print out all inconsistencies so the on-call or dev can take appropriate action for all inconsistencies at once.

Comment (snalli) [reply ambry-commons/src/main/java/com/github/ambry/clustermap/HelixVcrUtil.java:328]: the job is set to run every 1 hr

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/ConnectionStringBasedStorageClient.java (hunk 2 above)
Comment (snalli) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/azure/ConnectionStringBasedStorageClient.java:119]: ack

File: ambry-cloud/src/main/java/com/github/ambry/cloud/VcrReplicationManager.java (hunk 3 above)
Comment (snalli) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/VcrReplicationManager.java:381]: i merged container-deprecation code with regular cloud-compaction. the cloud-compaction scans all blobs and checks if any belongs to a deprecated container and deletes them in its regular flow.

imo, this separate job for deprecated containers was horrible idea.

File: ambry-commons/src/main/java/com/github/ambry/clustermap/HelixVcrUtil.java (hunk 1 above)
Comment (justinlin-linkedin) [reply ambry-commons/src/main/java/com/github/ambry/clustermap/HelixVcrUtil.java:328]: But we have an schedule task to sync them up, so it doesn't really need human intervention. If 1 hour is too long, we can set it to 10 minutes.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/VcrReplicationManager.java (hunk 3 above)
Comment (justinlin-linkedin) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/VcrReplicationManager.java:381]: that makes sense.

**Review by justinlin-linkedin — COMMENTED** [review]
Aside from the HelixVcrUtils print out all inconsistency, I have no issue with this PR

File: ambry-commons/src/main/java/com/github/ambry/clustermap/HelixVcrUtil.java (hunk 1 above)
Comment (snalli) [reply ambry-commons/src/main/java/com/github/ambry/clustermap/HelixVcrUtil.java:328]: changed to return on first inconsistency. but this whole code around helix sync-update task looks bad. Why do we have two separate tasks to check for sync and update ? They could be one job. Secondly, why aren't these jobs scheduled in the constructor ? these are being scheduled in helix context which is just horrible.

Comment (justinlin-linkedin) [reply ambry-commons/src/main/java/com/github/ambry/clustermap/HelixVcrUtil.java]: the issue is that we only want the host that have partition 0 to do the sync task so we don't run into any concurrency issues.

Comment (justinlin-linkedin) [reply ambry-commons/src/main/java/com/github/ambry/clustermap/HelixVcrUtil.java]: and separation of check and update task is indeed really weird.

**Review by justinlin-linkedin — APPROVED** [review]

**Pull request merged**